File: apps/android/app/src/main/java/com/example/creditcard/data/CardFutureFields.kt
package com.example.creditcard.data

/** Keep opaque fields from newer clients; never sync local UI state or legacy identities. */
internal object CardFutureFields {
    private val transient = setOf("showCardNumber", "showCVV", "countryRowSpan", "showCountry", "bankRowSpan", "showBank", "limitRowSpan", "showLimit", "lastTimeRowSpan", "showLastTime", "cardId", "uuid", "legacyId", "annualFeeDate", "extraFields", "constructor", "prototype")
    fun allowed(name: String) = !name.startsWith("_") && name !in transient
}

File: apps/web/src/utils/cardFutureFields.js
// Forward fields are opaque JSON, not editable UI or a new sync protocol.
const transient = new Set(["showCardNumber", "showCVV", "countryRowSpan", "showCountry", "bankRowSpan", "showBank", "limitRowSpan", "showLimit", "lastTimeRowSpan", "showLastTime", "cardId", "uuid", "legacyId", "annualFeeDate", "extraFields", "constructor", "prototype"])
export const allowedCardField = name => !name.startsWith('_') && !transient.has(name)
export const futureCardFields = (value, known) => Object.fromEntries(
  Object.entries(value).filter(([name]) => !known.has(name) && allowedCardField(name))
)
